### DISCUSSION TITLE
[React] Is there any way to avoid `overwrite` when a certain character is entered?

### DISCUSSION THREAD

**Discussion by DeltaSPb** [uNmAnNeR/imaskjs#1014] [Q&A] [answered]
Hi!

I'm using `react-imask` in my project and I have a question about `overwrite` behavior.
My mask for the input (_[codesandbox link](https://codesandbox.io/p/sandbox/masked-input-s6kpl3?file=%2Fsrc%2Findex.tsx%3A10%2C12&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cltve4g4l00063b6jz3sz63gj%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cltve4g4l00023b6jgkqzf84r%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cltve4g4l00033b6j7nrtdo32%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cltve4g4l00053b6jicsvk37u%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60.04095962551199%252C39.95904037448801%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cltve4g4l00023b6jgkqzf84r%2522%253A%257B%2522id%2522%253A%2522cltve4g4l00023b6jgkqzf84r%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cltve4g4l00053b6jicsvk37u%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltve4g4l00043b6jk36w1111%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cltve4g4l00053b6jicsvk37u%2522%252C%2522activeTabId%2522%253A%2522cltve4g4l00043b6jk36w1111%2522%257D%252C%2522cltve4g4l00033b6j7nrtdo32%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cltve4g4l00033b6j7nrtdo32%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A10%257D)_) consists of three groups and I need `overwrite={true}` behavior except for one case. If an input already has some value, but I need to insert an arithmetic character at the beginning of an existing string - inserting the character will overwrite the subsequent character:

https://github.com/uNmAnNeR/imaskjs/assets/32739170/d63ce5f8-def5-44e1-ab37-c72e9008a29b

Actual behavior:
`123° 45′ 67.8900″ => -23° 45′ 67.8900″`

Desired behavior:
`123° 45′ 67.8900″ => -123° 45′ 67.8900″`

Is there an easy way to make this work so that the `overwrite` doesn't trigger for only the case I described?

**Comment by uNmAnNeR** [comment, marked as answer]
hi @DeltaSPb,
well, it's a hard one.

I am going to have the ability to control `overwrite` from child masks in v8.
For now maybe custom dispatch could help.
Or toggle `overwrite` in `prepareChar` depending on the char. It's a dirty hack but it might help.
```jsx
<IMaskInput
  mask={mask}
  overwrite={true}
  placeholder={placeholder}
  placeholderChar={placeholderChar}
  type={"text"}
  inputRef={inputRef}
  value={value}
  autoComplete={"off"}
  onAccept={onChange}
  prepareChar={(char, masked) => {
    masked.overwrite = Boolean(masked.value) || char !== "-";
    return char;
  }}
  blocks={{
    DD: [{ mask: "0[00]" }, { mask: "-0[00]" }, { mask: "+0[00]" }],
    MM: { mask: "0[0]" },
    SS: { mask: "0[0]" },
    MS: { mask: "0[000]" },
  }}
/>
```

**Comment by DeltaSPb** [comment]
@uNmAnNeR 

Oh, thank you. It really works!
The only problem is that if the value in the first group is less than the maximum allowed (_in my case - 3 numeric characters_), when I add the `-` character, the input caret jumps to the end:


https://github.com/uNmAnNeR/imaskjs/assets/32739170/420aabeb-bf6a-4f4d-954c-9c35014bb193



Is it possible to avoid this behavior?

**Reply by uNmAnNeR** [reply]
@DeltaSPb 

```tsx
// avoid recreation
function prepareChar (char, masked) {
  masked.overwrite = Boolean(masked.value) || char !== "-";
  return char;
}

<IMaskInput
  mask={mask}
  // overwrite={true} <-- remove this to skip change detection
  placeholder={placeholder}
  placeholderChar={placeholderChar}
  type={"text"}
  inputRef={inputRef}
  value={value}
  autoComplete={"off"}
  onAccept={onChange}
  prepareChar={prepareChar}
  blocks={{
    DD: [{ mask: "0[00]" }, { mask: "-0[00]" }, { mask: "+0[00]" }],
    MM: { mask: "0[0]" },
    SS: { mask: "0[0]" },
    MS: { mask: "0[000]" },
  }}
/>
```

**Reply by DeltaSPb** [reply]
@uNmAnNeR Thanks a lot!
